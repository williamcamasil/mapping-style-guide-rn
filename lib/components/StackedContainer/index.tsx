import React, { PropsWithChildren, FunctionComponent } from 'react';
import {
  SafeAreaView, ScrollView, StyleProp, ViewStyle,
} from 'react-native';

import { useViewStyles } from '../../hooks';
import { AppThemeType, withTheme } from '../../theme';
import Container, { ContainerPropsType } from '../Container';
import KeyboardAvoidingWrap from '../KeyboardAvoidingWrap';
import Spacer from '../Spacer';

type StackedContainerPropsType = ContainerPropsType & PropsWithChildren & {
  theme: AppThemeType;
  containerStyle?: StyleProp<ViewStyle>;
  headerContent?: React.ReactNode;
  topContent?: React.ReactNode;
  middleContent?: React.ReactNode;
  bottomContent?: React.ReactNode;
  padding?: number;
};

const StackedContainer: FunctionComponent<StackedContainerPropsType> = ({
  theme,
  headerContent: headerChildren,
  topContent: topChildren,
  middleContent: middleChildren,
  bottomContent: bottomChildren,
  padding = theme.spacings.sLarge,
  ...others
}) => {

  const scrollViewContentStyle = useViewStyles(() => [
    {
      flexGrow: 1,
      paddingHorizontal: padding,
      justifyContent: 'space-between',
    },
  ], [padding]);

  const renderChildren = (children?: React.ReactNode) => {
    if (!children) return null;
    return (
      <SafeAreaView>
        {children}
        <Spacer size={padding} />
      </SafeAreaView>
    );
  };

  return (
    <Container {...others}>
      {headerChildren}
      <KeyboardAvoidingWrap>
        <ScrollView
          contentContainerStyle={scrollViewContentStyle}
          keyboardShouldPersistTaps="handled"
        >
          {renderChildren(topChildren)}
          {renderChildren(middleChildren)}
          {renderChildren(bottomChildren)}
        </ScrollView>
      </KeyboardAvoidingWrap>
    </Container>
  );
};

export default withTheme(StackedContainer);
