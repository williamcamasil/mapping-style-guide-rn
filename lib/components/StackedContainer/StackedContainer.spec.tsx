
import React from 'react';
import { Text, View } from 'react-native';

import { render } from '@testing-library/react-native';

import StackedContainer from '.';
import { ThemeProvider } from '../../theme';

describe('StackedContainer', () => {
  it('should render full stack', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          headerContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          topContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          middleContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          bottomContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only header', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          headerContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only top', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          topContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only middle', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          middleContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only bottom', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          bottomContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only top and bottom', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          topContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          bottomContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });

  it('should render only header, top and bottom', () => {
    const result = render(
      <ThemeProvider>
        <StackedContainer
          headerContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          topContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
          bottomContent={(
            <View>
              <Text>
                StackedContainer
              </Text>
            </View>
          )}
        />
      </ThemeProvider>,
    );

    expect(result).toMatchSnapshot();
  });
});
