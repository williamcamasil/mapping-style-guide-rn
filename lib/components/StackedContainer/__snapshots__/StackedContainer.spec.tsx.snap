// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedContainer should render full stack 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View>
    <Text>
      StackedContainer
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only bottom 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only header 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View>
    <Text>
      StackedContainer
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only header, top and bottom 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View>
    <Text>
      StackedContainer
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only middle 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only top 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`StackedContainer should render only top and bottom 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-between",
          "paddingHorizontal": 28,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
        <RCTSafeAreaView>
          <View>
            <Text>
              StackedContainer
            </Text>
          </View>
          <Text
            style={
              {
                "height": 28,
                "width": 28,
              }
            }
          />
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
